editors/code/src/scopes.ts

```diff
@@ -0,0 +1,146 @@
+    return (
+        extension.extensionKind === vscode.ExtensionKind.UI &&
+        extension.packageJSON.contributes &&
+        extension.packageJSON.contributes.themes
+    );
+}
+
+// Find current theme on disk
+function loadThemeNamed(themeName: string) {
+    const themePaths = vscode.extensions.all
+        .filter(filterThemeExtensions)
+        .reduce((list, extension) => {
+            return extension.packageJSON.contributes.themes
+                .filter(
+                    (element: any) =>
+                        (element.id || element.label) === themeName
+                )
+                .map((element: any) =>
+                    path.join(extension.extensionPath, element.path)
+                )
+                .concat(list);
+        }, Array<string>());
+
+    themePaths.forEach(loadThemeFile);
+
+    const tokenColorCustomizations: [any] = [
+        vscode.workspace
+            .getConfiguration('editor')
+            .get('tokenColorCustomizations')
+    ];
+
+    tokenColorCustomizations
+        .filter(custom => custom && custom.textMateRules)
+        .map(custom => custom.textMateRules)
+        .forEach(loadColors);
+}
+
+function loadThemeFile(themePath: string) {
+    const themeContent = [themePath]
+        .filter(isFile)
+        .map(readFileText)
+        .map(parseJSON)
+        .filter(theme => theme);
+
+    themeContent
+        .filter(theme => theme.tokenColors)
+        .map(theme => theme.tokenColors)
+        .forEach(loadColors);
+
+    themeContent
+        .filter(theme => theme.include)
+        .map(theme => path.join(path.dirname(themePath), theme.include))
+        .forEach(loadThemeFile);
+}
+
+function mergeRuleSettings(
+    defaultSetting: TextMateRuleSettings | undefined,
+    override: TextMateRuleSettings
+): TextMateRuleSettings {
+    if (defaultSetting === undefined) {
+        return override;
+    }
+    const mergedRule = defaultSetting;
+
+    mergedRule.background = override.background || defaultSetting.background;
+    mergedRule.foreground = override.foreground || defaultSetting.foreground;
+    mergedRule.fontStyle = override.fontStyle || defaultSetting.foreground;
+
+    return mergedRule;
+}
+
+function updateRules(
+    scope: string,
+    updatedSettings: TextMateRuleSettings
+): void {
+    [rules.get(scope)]
+        .map(settings => mergeRuleSettings(settings, updatedSettings))
+        .forEach(settings => rules.set(scope, settings));
+}
+
+function loadColors(textMateRules: TextMateRule[]): void {
+    textMateRules.forEach(rule => {
+        if (typeof rule.scope === 'string') {
+            updateRules(rule.scope, rule.settings);
+        } else if (rule.scope instanceof Array) {
+            rule.scope.forEach(scope => updateRules(scope, rule.settings));
+        }
+    });
+}
+
+function isFile(filePath: string): boolean {
+    return [filePath].map(fs.statSync).every(stat => stat.isFile());
```

Comment:
This `[x].map(foo).every(bar)` idiom (as opposed to `bar(foo(x))` is rather new to me. There are a couple more instances of it.

Also, can we make the theme loading async? Is it worth it?

Reply:
@lnicola It was originally async and it is worth going that way, but it would require changes to the over all code unrelated to this PR. We would have to change other parts into async *before* doing it for the theme loading. It was originally async but didn't work since for instance the config stuff isn't. 

I don't mind updating everything to be async if you guys are up for it. VSCode API itself is async.

Reply:
> This `[x].map(foo).every(bar)` idiom (as opposed to `bar(foo(x))` is rather new to me. There are a couple more instances of it.
> 
> Also, can we make the theme loading async? Is it worth it?

I prefer happy path railway instead of nested functions as the trail of thoughts become clearer and is easier to change as I just need to focus on input -> output. 

https://github.com/rust-analyzer/rust-analyzer/blob/eec68e6f451ee7675ce2cb29b696df091e6aed13/crates/ra_ide_api/src/inlay_hints.rs#L53-L59

Reply:
> We would have to change other parts into async before doing it for the theme loading.

I see what you mean. We can make it async later.

Reply:
I don't mind doing it, I was half way there, before realizing you might not want it and scrapped it, but it would have to be a separate PR branched from master.

Reply:
Agreed.